server/routes/bookings/cancelVisitController.ts

```diff
@@ -0,0 +1,112 @@
+import type { RequestHandler } from 'express'
+import { Meta, ValidationChain, matchedData, param, body, validationResult } from 'express-validator'
+import { BookerService, VisitService } from '../../services'
+import paths from '../../constants/paths'
+
+export default class CancelVisitController {
+  public constructor(
+    private readonly bookerService: BookerService,
+    private readonly visitService: VisitService,
+  ) {}
+
+  public confirmCancelView(): RequestHandler {
+    return async (req, res) => {
+      const { booker, bookings } = req.session
+      const { visits } = bookings
+
+      const errors = validationResult(req)
+      if (!errors.isEmpty() || bookings.type !== 'future') {
+        return res.redirect(paths.BOOKINGS.HOME)
+      }
+
+      const prisoner = booker.prisoners?.[0]
+        ? booker.prisoners?.[0]
+        : (await this.bookerService.getPrisoners(booker.reference))?.[0]
+
+      const { visitDisplayId } = matchedData<{ visitDisplayId: string }>(req)
+
+      const visit = visits.find(v => v.visitDisplayId === visitDisplayId)
+
+      return res.render('pages/bookings/cancel/cancel', {
+        errors: req.flash('errors'),
+        booker,
+        prisoner,
+        visit,
+        visitDisplayId,
+        showServiceNav: true,
+      })
+    }
+  }
+
+  public submit(): RequestHandler {
+    return async (req, res, next) => {
+      const { booker, bookings } = req.session
+      const { visits } = bookings
+      const { cancelBooking } = req.body
+
+      const { visitDisplayId } = matchedData<{ visitDisplayId: string }>(req)
+
+      const errors = validationResult(req)
+      if (!errors.isEmpty()) {
+        req.flash('errors', errors.array())
+
+        if (!visitDisplayId) {
+          return res.redirect(paths.BOOKINGS.HOME)
+        }
+
+        req.flash('formValues', matchedData(req, { onlyValidData: false }))
```

Comment:
Might not need this line 🤔. No free text fields or similar to pre-populate: so either it was the expected `yes|no` or it's something unexpected we don't want to hang on to?

Reply:
Also, not actually reading `formValues` from flash in the GET method - so it should go. Otherwise it'll be left 'orphaned' in flash and potentially picked up elsewhere.